### PR TITLE
Dynamo support for autograd.Function w/ once_differentiable

### LINKED ISSUE #106893: [autograd.Function] torch.compile w/ once_differentiable leads to opaque graph break
### 🐛 Describe the bug

Some internal models use once_differentiable. The error message is very opaque:
```py
import torch
from typing import *
from torch.autograd.function import once_differentiable


class ScaleGradient(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    @once_differentiable
    def backward(ctx, grad):
        return grad * 0.5

x = torch.randn(3, requires_grad=True)
def f(x):
    return ScaleGradient.apply(x)
output = torch.compile(f, backend='eager', fullgraph=True)(x)
```
returns
```
Unsupported: inline in skipfiles: ScaleGradient.backward  | wrapper /raid/rzou/pt/debug-cpu2/torch/aut
ograd/function.py

from user code:
   File "<ipython-input-20-149cb1010897>", line 18, in f
    return ScaleGradient.apply(x)
  File "/raid/rzou/pt/debug-cpu2/torch/_dynamo/variables/misc.py", line 241, in trampoline_autograd_bw
d
    return fn_cls.backward(*args, **kwargs)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

```

cc @ezyang @gchanan @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov 

### Error logs

_No response_

### Minified repro

_No response_

### Versions

main

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108686

Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR adds a mechanism to tell Dynamo
to inline a function, no matter if it is included in skipfiles.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D49064185](https://our.internmc.facebook.com/intern/diff/D49064185)

### CODE REVIEW

File: torch/_dynamo/symbolic_convert.py [hunk 1]
```diff
@@ -2207,13 +2207,30 @@ def check_inlineable(func):
...
+                isinstance(func, UserFunctionVariable) and
+                func.get_filename().endswith('torch/autograd/function.py') and
+                not func.fn.__module__.startswith('torch.autograd.function')
+            ):
+                return
+            # Function wrapped by torch.autograd.function.once_differentiable is known sound
+            if (
+                isinstance(func, NestedUserFunctionVariable) and
+                func.get_filename().endswith('torch/autograd/function.py') and
+                func.fn_name.value == 'once_differentiable.<locals>.wrapper.<locals>.<genexpr>'
```
Comment (zou3519) [inline torch/_dynamo/symbolic_convert.py:2230]: Seems super hacky. I can try the "decorator to mark a function to always be inlined" idea if we think this is too hacky.

Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:2230]: Yeah as this is, it doesn't spark joy.

One thing we can do is like `wrapper.once_differentiable = True` and then just hasattr check that field? Still gross, but beats this value check on string (which is not safe anyway, if someone refactors once_differentiable lol)

What was wrong with inlining btw?

Comment (zou3519) [reply torch/_dynamo/symbolic_convert.py:2230]: > What was wrong with inlining btw?

Before this PR, Dynamo attempts to inline the function produced by once_differentiable. Then, it fails in `InliningInstructionTranslator.check_inlineable` and complains that the function being inlined is in skipfiles.

File: torch/_dynamo_utils.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+"""
+Helper utilities for Dynamo that do not require a full import of torch._dynamo.
+The motivation for this file is to avoid circular dependencies.
+"""
+
+DYNAMO_FORCE_INLINE: Any = weakref.WeakKeyDictionary()
+
+
+def compiler_force_inline(f):
```
Comment (voznesenskym) [inline torch/_dynamo_utils.py:12]: nit: drop compiler from the fn names if its under _dynamo_utils?

File: torch/_dynamo_utils.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    This matters for lambdas and nested functions: each new instance
+    has a different __code__.
+
+    This API is a short-term hack to avoid design flaws in SKIPFILES:
+    there was not an easy way to mark a single API in a file that is
+    in SKIPFILES for inlining. We should rework the SKIPFILES design
+    (and this API).
+    """
+    DYNAMO_FORCE_INLINE[f.__code__] = True
```
Comment (voznesenskym) [inline torch/_dynamo_utils.py:24]: nit assert function type? even some `callable` passing objects like functools.partial won't have a __code__

File: torch/_dynamo_utils.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return f
+
+
+def compiler_should_force_inline(func):
+    return func.get_code() in DYNAMO_FORCE_INLINE
```
Comment (voznesenskym) [inline torch/_dynamo_utils.py:29]: this func is a VT, right? Let's assert that its a known VT type, and maybe move this to regular dynamo utils?

File: torch/autograd/function.py [hunk 5]
```diff
@@ -550,7 +551,9 @@ def apply(cls, *args, **kwargs):
 
 
 def once_differentiable(fn):
+    @compiler_force_inline
     @functools.wraps(fn)
+    @compiler_force_inline
```
Comment (voznesenskym) [inline torch/autograd/function.py:556]: Comment please?

**Review by voznesenskym — APPROVED** [review]

File: torch/autograd/function.py [hunk 6]
```diff
@@ -550,7 +551,14 @@ def apply(cls, *args, **kwargs):
...
 
 def once_differentiable(fn):
+    # In order for the function returned by once_differentiable to work with Dynamo,
+    # Dynamo needs to be able to inline through its return to prove safety.
+    # However, Dynamo thinks the function returned by once_differentiable is in a
+    # SKIPFILES, because this file is in SKIPFILES, which prevents inlining. We apply
+    # the following decorator (once per function) to force inlines.
+    @torch._dynamo_utils.force_inline
     @functools.wraps(fn)
+    @torch._dynamo_utils.force_inline
```
Comment (xw285cornell) [inline torch/autograd/function.py:561]: this line is duplicated?

Comment (zou3519) [reply torch/autograd/function.py:561]: This is intentional, we need to force_inline on `wrapper` and force_inline on the function produced by functools.wraps

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49064185).
